Block Entities article (#238)
FabricMC/fabric-docs: develop/blocks/block-entities.md

---
title: Block Entities
description: Learn how to create block entities for your custom blocks.
authors:
  - natri0
---

# Block Entities {#block-entities}

Block entities are a way to store additional data for a block, that is not part of the block state: inventory contents, custom name and so on.

> Vanilla Minecraft uses block entities for blocks like chests, furnaces, and command blocks.

Reviewer: ```suggestion
Minecraft uses block entities for blocks like chests, furnaces, and command blocks.
```

> In this article, we will create a block that counts how many times it has been right-clicked.

Reviewer: ```suggestion
As an example, we will create a block that counts how many times it has been right-clicked.
```

## Creating the Block Entity {#creating-the-block-entity}

> To make Minecraft "see" the new block entity, we need to create a block entity type. This is done by extending the `BlockEntity` class and registering it in the `ModBlockEntities` class, just like we did with the block.

Reviewer: ```suggestion
To make Minecraft recognize and load the new block entities, we need to create a block entity type. This is done by extending the `BlockEntity` class and registering it a new `ModBlockEntities` class.
```

@[code transcludeWith=:::1](@/reference/latest/src/main/java/com/example/docs/block/entity/custom/CounterBlockEntity.java)

> Registering a Block Entity yields a `BlockEntityType` like the `COUNTER_BLOCK_ENTITY` we've used above:

Reviewer: ```suggestion
Registering a `BlockEntity` yields a `BlockEntityType` like the `COUNTER_BLOCK_ENTITY` we've used above:
```

@[code transcludeWith=:::4](@/reference/latest/src/main/java/com/example/docs/block/entity/custom/CounterBlockEntity.java)

Now, if we save and reload the game, the counter block should continue from where it left off when saved.

## Tickers {#tickers}

The `BlockEntityProvider` interface also defines a method called `getTicker`, which can be used to run code every tick for each instance of the block. We can implement that by creating a static method that will be used as the `BlockEntityTicker`:

The `getTicker` method should also check if the passed `BlockEntityType` is the same as the one we're using, and if it is, return the function that will be called every tick. Thankfully, there is a utility function that does the check in `BlockWithEntity`:

@[code transcludeWith=:::3](@/reference/latest/src/main/java/com/example/docs/block/custom/CounterBlock.java)

`CounterBlockEntity::tick` is a reference to the static method `tick` we should create in the `CounterBlockEntity` class. Structuring it like this is not required, but it's a good practice to keep the code clean and organized.

> Let's say we want to make it so that the counter can only be incremented once every 10 ticks (2 times a second). We can do this by adding a `ticksSinceLast` field to the `CounterBlockEntity` class, and increasing it every tick:

Reviewer: Maybe mention something like "Don't forget to serialize and deserialize this new field!"
